src/main/core/manager.rs

```diff
@@ -255,30 +255,41 @@ impl<'a> Manager<'a> {
         let dns = unsafe { c::dns_new() };
         assert!(!dns.is_null());
 
-        for (i, info) in manager_config.hosts.iter().enumerate() {
-            let id = HostId::from(u32::try_from(i).unwrap());
-            let std::net::IpAddr::V4(addr) = info.ip_addr.unwrap() else {
-                unreachable!("IPv6 not supported");
-            };
-
-            let hostname = info.name.clone();
-            let chostname = CString::new(&*hostname).unwrap();
-            let caddr = u32::from(addr).to_be();
+        // Assign the host id only once to guarantee it stays associated with its host.
+        let host_init: Vec<(&HostInfo, HostId)> = manager_config
+            .hosts
+            .iter()
+            .enumerate()
+            .map(|(i, info)| {
+                // Set up the host identity.
+                let id = HostId::from(u32::try_from(i).unwrap());
+                let std::net::IpAddr::V4(addr) = info.ip_addr.unwrap() else {
+                    unreachable!("IPv6 not supported");
+                };
```

Comment:
(Optional) Maybe split these up into two steps? I think it's a little nicer for the functional style methods like `map` to not have side effects, but I'm sure there are places in shadow I've done this.

```rust
let host_init: Vec<(&HostInfo, HostId)> = manager_config
    .hosts
    .iter()
    .enumerate()
    .map(|(i, info)| (info, HostId::from(u32::try_from(i).unwrap())));

for (info, id) in &host_init {
    let std::net::IpAddr::V4(addr) = info.ip_addr.unwrap() else {
        unreachable!("IPv6 not supported");
    };
    ...
    unsafe { c::dns_register(dns, id, chostname.as_ptr(), caddr) };
}
```

Reply:
Done in e1ba7cba601d1d329c1ea3da3bf365b66f63ba12